packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,436 @@
+    this.alignment,
+    this.repeat: ImageRepeat.noRepeat,
+    this.centerSlice,
+  }) : assert(placeholder != null),
+      assert(image != null),
+      placeholder = new MemoryImage(placeholder, scale: scale),
+      image = new NetworkImage(image, scale: scale),
+      assert(fadeOutDuration != null),
+      assert(fadeOutCurve != null),
+      assert(fadeInDuration != null),
+      assert(fadeInCurve != null),
+      super(key: key);
+
+  /// Creates a widget that uses a placeholder image stored in an asset bundle
+  /// while loading the final image from the network.
+  ///
+  /// [placeholder] is the key of the image in the asset bundle.
+  ///
+  /// [image] is the URL of the final image.
+  ///
+  /// The [placeholder], [image], [fadeOutDuration], [fadeOutCurve],
+  /// [fadeInDuration] and [fadeInCurve] arguments must not be null.
+  ///
+  /// See also:
+  ///
+  ///  * [new Image.asset], which has more details about loading images from
+  ///    asset bundles.
+  ///  * [new Image.network], which has more details about loading images from
+  ///    the network.
+  FadeInImage.assetNetwork({
+    Key key,
+    @required String placeholder,
+    @required String image,
+    AssetBundle bundle,
+    double scale: 1.0,
+    this.fadeOutDuration: const Duration(milliseconds: 300),
+    this.fadeOutCurve: Curves.easeOut,
+    this.fadeInDuration: const Duration(milliseconds: 700),
+    this.fadeInCurve: Curves.easeIn,
+    this.width,
+    this.height,
+    this.fit,
+    this.alignment,
+    this.repeat: ImageRepeat.noRepeat,
+    this.centerSlice,
+  }) : assert(placeholder != null),
+       assert(image != null),
+       placeholder = scale != null
+         ? new ExactAssetImage(placeholder, bundle: bundle, scale: scale)
+         : new AssetImage(placeholder, bundle: bundle),
+       image = new NetworkImage(image, scale: scale),
+       assert(fadeOutDuration != null),
+       assert(fadeOutCurve != null),
+       assert(fadeInDuration != null),
+       assert(fadeInCurve != null),
+       super(key: key);
+
+  /// Image displayed while the target [image] is loading.
+  final ImageProvider placeholder;
+
+  /// The target image that is displayed.
+  final ImageProvider image;
+
+  /// The duration of the fade-out animation for the [placeholder].
+  final Duration fadeOutDuration;
+
+  /// The curve of the fade-out animation for the [placeholder].
+  final Curve fadeOutCurve;
+
+  /// The duration of the fade-in animation for the [image].
+  final Duration fadeInDuration;
+
+  /// The curve of the fade-in animation for the [image].
+  final Curve fadeInCurve;
+
+  /// If non-null, require the image to have this width.
+  ///
+  /// If null, the image will pick a size that best preserves its intrinsic
+  /// aspect ratio.
+  final double width;
+
+  /// If non-null, require the image to have this height.
+  ///
+  /// If null, the image will pick a size that best preserves its intrinsic
+  /// aspect ratio.
+  final double height;
```

Comment:
here and under width, maybe mention that if the placeholder and target images don't have the same aspect ratio, not specifying them will result in a sudden change in the layout? (assuming that's true?)